### ISSUE TITLE
Introduce concat or bytes.concat

### ISSUE THREAD

**Issue by axic** [ethereum/solidity#10903]
This came up frequently and has been recently discussed under #9829. One of the uses of `abi.encodePacked` is for it to perform concatenation.

Many EIPs use snippets to explain the rules and many libraries rely on `abi.encodePacked`. The first impression is that they actually rely on the weird "encoding" it performs, but a closer look reveals that in many cases they only rely on the concatenation feature. In many cases the concatenation is used to add a prefix prescribed by [EIP-191](https://eips.ethereum.org/EIPS/eip-191). An example for actual data encoding is [EIP-712](https://eips.ethereum.org/EIPS/eip-712#definition-of-typed-structured-data-%F0%9D%95%8A). 

While it is a bit unfortunate that standards (aka EIPs) rely on Solidity snippets, it is something we likely can not change, but we can improve the language so that confusion is also minimised in those standards. Having `concat` instead of `encodePacked` I think would go towards that. (I believe this confusion lead to [libraries adding support](https://docs.ethers.io/v5/api/utils/hashing/#utils--solidity-hashing) for "encodePacked".)

I propose to introduce `concat` which takes a variable number of `bytes memory` as an input and returns a single `bytes memory`.

Question A: should it also allow fixed bytes (i.e. `bytesNN`) as an inputs, which are tightly packed and not zero padded? Or should we in parallel [finally implement conversion between fixed and variable bytes types](https://github.com/ethereum/solidity/issues/9170)?

Question B: should we start adding functions as members of types, i.e. `bytes.concat()`? If yes, is that a breaking change? I guess it is, because libraries could extend those types already.

**Comment by chriseth** [comment]
I don't think a global `concat` would be nice. Maybe we should start introducing "built-in" files that you can the use as
```
import { concat } from utils;
```
(note that there are no quotes around `utils`)

**Comment by hrkrshnn** [comment]
Does `concat` really have to take a variable number of `bytes memory`? I think we should just make it have two arguments. This would allow us to move it into the "standard library" in the future.

**Comment by ekpyron** [comment]
> Does `concat` really have to take a variable number of `bytes memory`? I think we should just make it have two arguments. This would allow us to move it into the "standard library" in the future.

Could the optimizer deal with that properly (now or ever) :-)? A ``concat`` with a variable number of arguments can first sum up all length and allocate a huge chunk of result memory once. Iterated ``concat``s with two arguments would allocate a whole bunch of temporary arrays in between, but all of them will be required until the last one was created. And the size of the last one is unknown until you created all of the ones up front - that seems like a pretty hard problem to reduce to a single allocation in the end.
Apart from that we should have variadic templates, then we could move it to the standard library either way ;-).

**Comment by hrkrshnn** [comment]
The uses for `abi.encodePacked` according to https://github.com/search?p=1&q=abi.encodepacked&type=Code is surprisingly mostly for a single argument!

**Comment by chriseth** [comment]
What about defining the `+` operator?

**Comment by axic** [comment]
Could also consider other operators for concatenation, such as `..`. (I just realised it is not used frequently, but in Lua, which I used for years 😬)

Also see: https://en.wikipedia.org/wiki/Concatenation

**Comment by axic** [comment]
Other operators were suggested:
- `++`
- `~`
- `||`

**Comment by axic** [comment]
Decision from call: `bytes.concat` with variadic arguments supporting fixed and variable length bytes.

We also discussed the option to have `bytesNN.concat`, but deferred decision. One use case people have is keccak hashing with `encodePacked`, because that has a memory wasting behaviour. There having `bytesNN.concat` would be beneficial. We concluded that we should attempt to improve memory handing for the `keccak256(bytes.concat())` (see #9046) case before coming back to `bytesNN.concat`.

**Comment by chriseth** [comment]
@mijovic what is still left to do here?

**Comment by mijovic** [comment]
I think I didn't close the issue because there was still an open point about allowing `bytesNN` as an argument of `bytes.concat()`. If we do not want to allow that, I think there is nothing else to be done here.
Also, returning fixed bytes type, for example `bytes32.concat()` but probably after finishing bytes->bytesNN (#11047) it is not relevant as we can convert it easily.

**Comment by chriseth** [comment]
I would say this is done.
